File: Antony-Ndungu/Arrays and Strings 1/reverse-sentence-in-place.go
package main

import (
	"fmt"
)

func main() {
	A := []byte{'t', 'h', 'i', 's', ' ', 'i', 's', ' ', 'g', 'o', 'o', 'd'}
	print(A, "Before")
	reverseSentence(A)
	print(A, "After")
}

func print(A []byte, description string) {
	fmt.Printf("%s: [", description)
	for i, c := range A {
		fmt.Printf("'%c'", c)
		if i == len(A)-1 {
			continue
		}
		fmt.Printf(",")
	}
	fmt.Printf("]\n")
}

func reverseArray(A []byte, startIndex int, endIndex int) {
	for i, j := startIndex, endIndex; i < j; i, j = i+1, j-1 {
		A[i], A[j] = A[j], A[i]
	}
}

func reverseSentence(A []byte) {
	// Reverse the characters in the sentence
	reverseArray(A, 0, len(A)-1)
	//Traverse the sentence, reversing each word in the sentence
	var startOfWord, endOfWord int
	for i := 0; i < len(A); i++ {
		if A[i] == ' ' {
			endOfWord = i - 1
			reverseArray(A, startOfWord, endOfWord)
			startOfWord = i + 1
			continue
		}
		if i == len(A)-1 {
			reverseArray(A, startOfWord, len(A)-1)
		}
	}
}

File: Antony-Ndungu/Arrays and Strings 1/reverse-sentence-in-place.c
#include <stdio.h>

void print(char *A, size_t length, char *description);
void swap(char *a, char *b);
void reverseArray(char *A, size_t startIndex, size_t endIndex);
void reverseSentence(char *A, size_t length);

int main(void){
    char A[] ={'t','h','i','s',' ','i','s',' ','g','o','o','d'};
    size_t length = sizeof(A)/sizeof(char);
    print(A, length, "Before");
    reverseSentence(A,length);
    print(A, length, "After");
}

void print(char *A, size_t length, char *description){
    printf("%s: [", description);
    for(size_t i = 0; i < length; i++){
        printf("'%c'", A[i]);
        if(i == length - 1)
            continue;
        printf(", ");
    }
    printf("]\n");
}

void swap(char *a, char *b){
    char temp = *a;
    *a = *b;
    *b = temp;
}

void reverseArray(char *A, size_t startIndex, size_t endIndex){
    for(size_t i = startIndex, j = endIndex; i < j; i++, j--)
        swap(A+i, A+j);
}

void reverseSentence(char *A, size_t length){
    //Reverse sentence
    reverseArray(A, 0, length -1);
    size_t startOfWord = 0, endOfWord = 0;
    // Traverse the sentence, reversing each word
    for(size_t i = 0; i < length; i++){
        if(A[i] == ' '){
            endOfWord = i - 1;
            reverseArray(A, startOfWord, endOfWord);
            startOfWord = i + 1;
            continue;
        }
        if(i == length - 1){
            reverseArray(A, startOfWord, length-1);
        }
    }
}